Lab #8 on Programming

MrDvD opened the pull request.

MrDvD committed:
chore: Init a lab

MrDvD committed:
feat: integrate Protobuf into collections (haven't tested yet)

MrDvD committed:
feat: Working gRPC /GetInfo endpoint

Remove old services related to Selector logic
Style-check of project

MrDvD committed:
feat: Implement /AddTicket endpoint

Fix error handling for /GetTickets

MrDvD committed:
feat: Implement /AddTicket endpoint

Add ShowByIdCommand at PublicExecutor

MrDvD committed:
feat: Add AuthInterceptor for login purposes

Add template for ExceptionInterceptor

MrDvD committed:
refactor: Remove Proxy interface and related strategies

Change of commands RemoveAt, RemoveLast -> ShowAt, ShowLast
Add AuthService template & related integrations for future
Clean-up of CachedCrudWorker

MrDvD committed:
feat: Integrate JWT into auth process

Remove unused service interfaces from 7th lab
Add UserService for authorization
Style-check of project

MrDvD committed:
feat: Implement /GetInfo, /UpdateTicket, /RemoveAt, /RemoveById

Fix at JDBC Ticket module (owner & timestamp)
Fix at /AddTicket, /RemoveLast

MrDvD committed:
chore: Restructure packages

Remove unused classes

MrDvD committed:
fix: Fix of JDBC methods for Ticket collections

Remove unnnecessary commands from executor
Update launch process in main() method

laoqiu233 commented on programming/8/server-api/src/main/proto/com/itmo/mrdvd/TicketService.proto:
```diff
@@ -0,0 +1,98 @@
+  // Updates the Ticket object in collection
+  rpc UpdateTicket(UpdateId) returns (google.protobuf.Empty) {}
+  // Removes the last added Ticket object from the collection
+  rpc RemoveLast(google.protobuf.Empty) returns (google.protobuf.Empty) {}
```
Это ок - но проблематично будет когда ты захочешь в будущем добавить новые поля к запросу или ответу

Хорошой практикой будет сделать пустой Message и использовать его, для ответа и запроса

laoqiu233 commented on programming/8/server-api/src/main/proto/com/itmo/mrdvd/TicketService.proto:
```diff
@@ -0,0 +1,98 @@
+  // Removes the Ticket object by its index from the collection 
+  rpc RemoveAt(IntId) returns (google.protobuf.Empty) {}
+  // Removes the Ticket object by its id from the collection
+  rpc RemoveById(LongId) returns (google.protobuf.Empty) {}
```
Это наверное плохая идея использовать такие примитивы в запросах и ответах

Это ОК для каждой ручки сделать свой message запроса - расширяемость в будущем у тебя будет 

Ну или как минимум какой нибудь `RemoveRequest`, но точне не сообщения с примитивами

laoqiu233 commented on programming/8/server-api/src/main/proto/com/itmo/mrdvd/TicketService.proto:
```diff
@@ -0,0 +1,98 @@
+  }
+}
+
+message Meta {
```
более человечный нейминг хочется увидеть, хотя бы `CollectionMeta`, просто так непонятно что это и для чего используется, когда в сообщении просто мапа

laoqiu233 commented on programming/8/server/src/main/java/com/itmo/mrdvd/service/interceptors/KeyFillContextInterceptor.java:
```diff
@@ -0,0 +1,19 @@
+          }
+          Object tokenValue = t.get(metaKey);
+          return Context.current().withValue(key, tokenValue);
+        });
```
вообще в жаве это лучше делать через абстрактный fillContext метод в fillContextInterceptor и тут наследоваться, как будто лямбду сложнее в перспетиве отлаживать

laoqiu233 commented on programming/8/server-api/src/main/proto/com/itmo/mrdvd/AuthService.proto:
```diff
@@ -0,0 +1,33 @@
+message AuthResponse {
+  AuthStatus status = 1;
+  optional AuthID id = 2;
+}
```
наверное тут больше подойдет какой нибудь oneof?

laoqiu233 reviewed (approved):
lgtm

смотрел только в окрестности с grpc, остальное кажется тоже ок

MrDvD committed:
feat: Launch project via docker-compose

MrDvD committed:
refactor: Change in Protobuf definitions for scalability

Rename interfaces for readability
Refactor of FillContextInterceptor
Merge endpoints of TicketService: /RemoveLast, /RemoveAt, /RemoveById -> /RemoveTicket

MrDvD commented:
@laoqiu233, внёс небольшие правки, также захотелось упаковать приложение в контейнер и поиграться с docker-compose (подглядел идею, понравилась задумка). Можешь глянуть, если будет время? Изменений немного, по diff двух последних коммитов можешь пройтись

laoqiu233 commented on programming/8/Dockerfile:
```diff
@@ -0,0 +1,7 @@
+COPY server/target/8-server-1.0-jar-with-dependencies.jar /app
+
+ENTRYPOINT ["java", "-jar", "8-server-1.0-jar-with-dependencies.jar"]
+CMD []
```
ну так не делается

обычно в dockerfile  прописываешь саму сборку, желательно в две стадии, одна билд стадия с образом jdk, вторая runtime которая будет использоваться при запуске чисто с jre

laoqiu233 commented on programming/8/compose.yaml:
```diff
@@ -0,0 +1,35 @@
+services:
```
обычно файл называют `docker-compose.yml`, а не просто `compose.yml`, ибо `docker-compose.yml` это дефолтное название файла, который подхватывается

MrDvD replied:
Когда выбирал имя файла, я вычитал, что `docker-compose.yml` уже устарел, и по умолчанию теперь подставляют `compose.yml` [(пруф)](https://docs.docker.com/compose/intro/compose-application-model/#the-compose-file)

laoqiu233 commented on programming/8/compose.yaml:
```diff
@@ -0,0 +1,35 @@
+services:
+  database:
+    image: "postgres"
```
версию укажи

MrDvD replied:
По умолчанию latest подставится, в чём проблема? Либо буду сидеть на одной версии и ждать, пока возникнет уязвимость, которой можно воспользоваться.

С другой стороны, не факт, что не появится новых уязвимостей в latest, поэтому желательно делать проверки инструментами osa/sca, если не путаю понятия. Для лабы, наверное, это будет излишним

laoqiu233 commented on programming/8/compose.yaml:
```diff
@@ -0,0 +1,35 @@
+    environment:
+      - POSTGRES_USER
+      - POSTGRES_PASSWORD
+    volumes:
+      - ./init.sql:/docker-entrypoint-initdb.d/init.sql
```
volume обычно делается чтоб персистить данные, посмотри где там хранятся в постгре данные и пропиши тут

laoqiu233 reviewed (approved):
минорные комменты, остальное lgtm

MrDvD committed:
fix: Split Dockerfile into two stages: build and runtime

MrDvD committed:
fix: Move validation logic in Jdbc components to the start

Suppress exception messages in Jdbc components (change to safe ones)
Remove old authorization logic in predicate lambda function of UpdateCommand

MrDvD committed:
chore: Add README.md

MrDvD committed:
fix: Persist postgres data from database

Update .gitignore
Added new content in README.md at 8th lab
